### DISCUSSION TITLE
scrollToTop in v3 ?

### DISCUSSION THREAD

**Discussion by andysay** [nuxt/nuxt#16223] [Questions]
v3 not support scrollToTop: true ?

**Comment by megasanjay** [comment]
Been trying to figure this one out as well. I don't want to set up manual scrolling for every page on our website either.

**Comment by BboyAkers** [comment]
Ended up creating a wrapper for `NuxtLink` to solve this temporarily 🙂 . I actually forked the repo and am looking to it because I was curious on why I wasn't able to use it! Will keep you updated!

My New Component:
NNuxtLink.vue
```vue
<script setup lang="ts">
const scrollTop = () => {
  window.scrollTo(0, 0)
}
</script>

<template>
  <NuxtLink @click="scrollTop"><slot /></NuxtLink>
</template>
```

**Comment by nachtfunke** [comment]
I am actually struggling with this as well. What I have so far from another thread, is trying to create `plugins/router.scrollBehaviour.js` with the contents:
```javascript
import { defineNuxtPlugin } from '#app';

export default defineNuxtPlugin((nuxtApp) => {
	nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
		return { x: 0, y: 0 };
	};
});
```
Which does nothing.

I also tried setting up a watcher for `$route` in `layouts/default.vue`:

```javascript
watch: {
        $route() {
            setTimeout( () => {
                window.scrollTo(0, 0);
            }, 100 );
       }
}
```
This also does not work. Scrolling just does not work at all it seems.

**Comment by aligzl** [comment]
I figure it out with fallowing code in app.vue

```
<script setup>

  const nuxtApp = useNuxtApp()

  nuxtApp.hook("page:finish", () => {
     window.scrollTo(0, 0)
  })
</script>
```

**Reply by lucassimines** [reply]
Thanks!!

**Reply by J4MSH1D** [reply]
Thanks

**Reply by bracketsdigital** [reply]
aligzl - you are a superstar!!

**Reply by arnosolo** [reply]
Thank you.

**Reply by trentbrew** [reply]
Thank you 💙

**Reply by nyomansunima** [reply]
It's work. Definitely i combine it using gsap and smooth scroll.

**Comment by megasanjay** [comment]
According to #1707 and #2330 I think going the plugin route is probably the best case. The only issue I found was that it seems to scroll to the top before the page transition starts. This can be slightly jarring but you can give it a try.

**Reply by jamaluddinrumi** [reply]
up

**Comment by jamaluddinrumi** [comment]
[UPDATED]

~i found the way to do this by using hooks (page:finish)~
i prefer to use router options 
and i also add a bit delay to it

i attached it on plugin 

here is my code below

```
// plugins/scrollToTop.client.js

import { defineNuxtPlugin } from "#app";

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve({ left: 0, top: 0, behaviour: "smooth" });
      }, 100);
    });
  };
});

```

i also set scroll-behaviour: smooth

```
// app.vue

<template>
  <div>
    <nuxt-layout>
      <nuxt-page />
    </nuxt-layout>
  </div>
</template>

<style>
@import "@/assets/css/tailwind.css";
@import "@/assets/css/inter.css";
@import "@/assets/css/app.css";

html {
  scroll-behavior: smooth; /* set scroll-behaviour to smooth here */
}
</style>

```

more about router options which is used by nuxt 3 are described on [here](https://router.vuejs.org/guide/advanced/scroll-behavior.html#delaying-the-scroll)

**Reply by olivervorasai** [reply]
How can these solutions be combined with page transitions?

For example, if you have page transitions defined like this
```css
.page-enter-active,
.page-leave-active {
    transition: all 2.5s;
}
.page-enter-from,
.page-leave-to {
    opacity: 0;
    transform: scale(0.8);
}
```

The scroll behavior will not wait 2.5 seconds for the page transitions to finish before scrolling.

**Comment by IzakJackson** [comment]
This kinda works - unfortunately, it also breaks the "color-theme" module.

**Reply by danielroe** [reply]
would you explain a bit more, and maybe open an issue on [the color-mode module](https://github.com/nuxt-community/color-mode-module)?

**Reply by IzakJackson** [reply]
Using this plugin:

```
import { defineNuxtPlugin } from "#app";

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve({ left: 0, top: 0, behaviour: "smooth" });
      }, 100);
    });
  };
});
```

When dark mode is enabled and I visit any page, it flashes dark mode briefly, then reverts to light mode (despite dark mode being enabled).

When I remove the plugin, dark mode functions as expected.

**Comment by misaon** [comment]
I prefer this solution:
```
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook('page:finish', () => {
    setTimeout(() => {
      window.scrollTo(0, 0)
    }, 300)
  })
})
```

**Comment by wobsoriano** [comment]
This is the implementation in Nuxt 2 https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-app/template/router.scrollBehavior.js

If anyone wants to convert it to a plugin

**Comment by martinrisseeuw** [comment]
For anyone looking for a solution, what i have done to mimic default browser behaviour.

1. Create a router.options.js file in an app folder in the root of your project
2. Add code below

```
// app/router.options.js
export default {
  scrollBehavior(to, from, savedPosition) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve({
          left: savedPosition?.left || 0,
          top: savedPosition?.top,
        });
      }, 0);
    });
  },
};
```

I'm currently running nuxt: `3.0.0-rc.3`

**Reply by tomblixt** [reply]
That's the best solution I have found so far!
I changed it a bit to scroll up. I also noticed that the promise wasn't necessary, so I removed it.
```
// app/router.options.js
export default {
  scrollBehavior(to, from, savedPosition) {
    return {
      left: 0,
      top: 0
    }
  }
}
```

**Reply by benoitdemaegdt** [reply]
Thanks. This works for me 👍 
You can get it even simpler : 
```
export default {
  scrollBehavior: () => ({ left: 0, top: 0 })
}
```

**Reply by martinrisseeuw** [reply]
With the code you share, you're missing the functionality of keeping the scroll position on the browser's back button. I wanted that functionality for the application, I'm working on. But when not necessary you can indeed remove the promise, timeout, and savedPosition.

**Reply by tomblixt** [reply]
You are right, I didn't know where the `savedPosition` was used for. I've now added `savedPosition?.top || 0` the the top property and now it works as I would expect. However I still don't understand where the promise is used for.

**Reply by vanling** [reply]
@martinrisseeuw THANKS <3

Used your code example :) Added 'smoothing' and the timeout was exactly what I needed for using this with page transitions :)

```
// app/router.options.js
export default {
  scrollBehavior(to, from, savedPosition) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve({
          behavior: 'smooth',
          top: savedPosition?.top || 0,
        })
      }, 440)
    })
  },
}
``

**Comment by shtse8** [comment]
or we can make it as plugin

```js
// plugins/router.ts
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          left: savedPosition?.left || 0,
          top: savedPosition?.top || 0,
        })
      }, 0)
    })
  }
})

```

**Reply by MartinX3** [reply]
`options` is unknown error.

**Reply by LuisSousaSilva** [reply]
This works as a good workaround. Thanks. But am i the only one who sees a ugly scroll flash? Anyway to get around it?

**Reply by benoitdemaegdt** [reply]
consider adding `behavior: 'smooth'`

```
// plugins/router.ts
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = (to, from, savedPosition) => {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          left: savedPosition?.left || 0,
          top: savedPosition?.top || 0,
          behavior: 'smooth'
        })
      }, 0)
    })
  }
})
```
It should be better 😉

**Reply by LuisSousaSilva** [reply]
thanks for the fast answer. I had tried that. But i feel that's not the default behavior in user experience.  Somehow, i still can't make the same seamless behavior i had with just plain old html+css+vanillaJS with a good CDN. Let's hope their implementation of scroll to top is somewhat better. https://v3.nuxtjs.org/migration/component-options#scrolltotop

**Comment by letoast** [comment]
Here's another one, this one checks for hash as well:

For me it works in `app/router.options.ts`

```js
// app/router.options.ts
import type { RouterConfig } from '@nuxt/schema'

export default <RouterConfig>{
    scrollBehavior: (to, from, savedPosition) => {
        return new Promise((resolve) => {
            setTimeout(() => {
                if (!to.hash) {
                    resolve({
                        left: 0,
                        top: 0,
                    })
                } else {
                    const top = (window.pageYOffset + document.querySelector(to.hash).getBoundingClientRect().top) || 0
                    resolve({
                        left: 0,
                        top,
                    })
                }
            }, 100)
        })
    }
}
```

**Reply by MartinX3** [reply]
Doesn't work. Also has deprecation warnings.

**Comment by ppodds** [comment]
Add this file to your plugins folder.

`plugins/scroll-top.ts`

```typescript
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook("page:finish", async (a) => {
    window.scrollTo({ top: 0 });
  });
});
```

It works fine with async page loading.

**Reply by Arturexe** [reply]
Shout out to this. best solution so far. also works great with page transitions if you add a delay:

```
export default defineNuxtPlugin((nuxtApp) => {
    nuxtApp.hook('page:finish', async (a) => {
        setTimeout(() => {
            window.scrollTo({ top: 0 })
        }, 300)
    })
})
```

**Comment by helloppl** [comment]
// call from a button click 
TippyTop(){
document.body.scroll({top:0,behavior:'smooth'})
}


// you're welcome

**Comment by MartinX3** [comment]
I created an issue ticket with a workaround
https://github.com/nuxt/bridge/issues/457

**Comment by bitmonti** [comment]
I use layouts and pages. In the script tag of my default layout i use the onUpdate hook.

`<script setup>onUpdated(() => window.scrollTo(0, 0));</script>`

**Comment by ram-you** [comment]
This one works for me : 
in app/router.options.js
```js
export default {
  scrollBehavior(to, from, savedPosition) {
    return new Promise((resolve, reject) => {
      if (to.hash) {
        setTimeout(() => {
          resolve({ el: to.hash, top: 60, behavior: 'smooth' })
        }, (!from || (to.path != from.path)) ? 500 : 1)
      } else {
        resolve({ top: 0, behavior: 'smooth' })
      }
    })
  }
}
```

**Comment by misaon** [comment]
Final solution for me

```ts
import { defineNuxtPlugin } from '#imports'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = async (to, from, savedPosition) => {
    let goTo

    if (to.hash) {
      goTo = {
        el: to.hash,
        behavior: 'smooth',
        top: 64,
      }
    } else if (savedPosition) {
      goTo = savedPosition
    } else {
      goTo = { top: 0, behavior: 'smooth' }
    }

    return new Promise((resolve) => {
      setTimeout(() => {
        resolve(goTo)
      }, 100)
    })
  }
})
```

**Comment by misaon** [comment]
Based on latest Nuxt edge:

```ts
import { useNuxtApp, defineNuxtPlugin } from '#imports'
import { nextTick } from 'vue'
import type { RouterScrollBehavior, RouteLocationNormalized } from 'vue-router'

type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.$router.options.scrollBehavior = async (to, from, savedPosition) => {
    let position: ScrollPosition = savedPosition || undefined

    if (!position && from && to && to.meta.scrollToTop !== false && _isDifferentRoute(from, to)) {
      position = { left: 0, top: 0 }
    }

    if (to.path === from.path) {
      if (from.hash && !to.hash) {
        return { left: 0, top: 0 }
      }

      if (to.hash) {
        return { el: to.hash, top: _getHashElementScrollMarginTop(to.hash) }
      }
    }

    return new Promise((resolve) => {
      nuxtApp.hooks.hook('page:finish', async () => {
        await nextTick()
        if (to.hash) {
          position = { el: to.hash, top: _getHashElementScrollMarginTop(to.hash) }
        }

        resolve(position)
      })
    })
  }
})

function _getHashElementScrollMarginTop(selector: string): number {
  const element = document.querySelector(selector)
  if (element) {
    return Number.parseFloat(getComputedStyle(element).scrollMarginTop)
  }

  return 0
}

function _isDifferentRoute(a: RouteLocationNormalized, b: RouteLocationNormalized): boolean {
  const samePageComponent = a.matched[0] === b.matched[0]

  if (!samePageComponent) {
    return true
  }

  return samePageComponent && JSON.stringify(a.params) !== JSON.stringify(b.params)
}
```

**Reply by vanling** [reply]
Woah! haha, Compared to the examples when the thread started hehehe this is amazing.

**Reply by quamsta** [reply]
~~This is awesome.. it doesn't seem to work for me in Firefox for some reason though :(~~ <- crossed this out.. it does work in Firefox, issue was on my end.

**Reply by chasebank** [reply]
This does indeed look awesome! But `savedPosition` is always showing as null for me. Any clue what could cause that?

**Reply by misaon** [reply]
For me nested pages not work correctly

**Reply by chasebank** [reply]
Looks like it actually is working, for back links 👍 https://stackblitz.com/edit/nuxt-starter-sqsaiv (not sure why vue-router breaks Stackblitz' back button, but it works when viewed in a new tab)

**Reply by ReaganM02** [reply]
Any updates on this one? I've tried all the solutions above but there is a strange behavior(see screen recording below). There are times that it works as expected but there are also times that when clicking one of the pagination controller it suddenly scrolls at the center. 
https://streamable.com/r0jqmh

**Reply by bitmonti** [reply]
Final solution for me on nuxt: 3.0.0-rc.13

Add page key
`<NuxtPage page-key="static" />`

Create plugin with Promise
plugins/scrollBehavior.js

```
export default defineNuxtPlugin((nuxt) => {
  nuxt.$router.options.scrollBehavior = (to, from, savedPosition) => {
    return new Promise((resolve) => {
      if (savedPosition) {
        resolve(savedPosition);
      } else {
        resolve({ top: 0 });
      }
    });
  };
});
```

**Reply by ReaganM02** [reply]
Thank you @bitmonti, your solution is working. One thing I noticed is that when you click on one of the `NuxtLink`, and when scrolling down fast, the scroll to top behavior breaks. It only stays at the bottom but I guess nobody would scroll down very fast on the page unless you are a developer or playing around with it.

**Reply by vanling** [reply]
So there are many reasons why a single script does not work for everyone.
it can be;
- Wrong usage of app.vue and/or layouts and multiple levels of <nuxtPage/>
- You are lazyloading data, and full page height is not yet 'calculated' for example: scroll tries to go to top:5000 and page is only 2500 high because not all data is loaded
- Issues with transitioning between layout
- You are using page transitions(this makes it custom for everyone else) page transitions can 'f' up the timings for scrolling.


Currently this is the one that is working _for me_ and it is completely based on @misaon his creations, just some minor alterations.
- I like to use smooth scrolling for back and forth on pages
- For saved positions im using a rather large timeout of 500ms before scrolling
- For normal page scrolls to top my timeout is a bit larger then the total page transition time. I use 230which is a bit more then the transition 110ms+delay _(and it can be my timings are of here now;) but i hope you get the idea)_

```
.page-enter-active, .page-leave-active { transition: 110ms all 110ms ease-in-out; }
.page-enter-active, .page-leave-active { transition: 110ms all 110ms ease-in-out; }
.page-enter-from, .page-leave-to {  opacity: 0; }
```

**For in `app/router.options.js`**
```
import { useNuxtApp } from '#imports'
export default {
  async scrollBehavior(to, from, savedPosition) {
    return new Promise(async (resolve, reject) => {
      const nuxtApp = useNuxtApp()

      if (to.hash) {
        setTimeout(
          () => {
            resolve({ el: to.hash, top: 60, behavior: 'smooth' })
          },
          !from || to.path != from.path ? 500 : 1
        )
      } else if (savedPosition) {
        nuxtApp.hooks.hook('page:finish', async () => {
          await nextTick()
          setTimeout(() => {
            resolve({ ...savedPosition, behavior: 'smooth' })
          }, 500)
        })
      } else {
        nuxtApp.hooks.hook('page:finish', async () => {
          await nextTick()
          setTimeout(() => {
            resolve({ top: 0, behavior: 'smooth' })
          }, 230)
        })
      }
    })
  },
}
```


**tl;dr** there is currently no 'one size fits all' to make this smooth. But thats the fun of this thing haha! fiddle around to make it work smooth for your app/site/setup.

**Comment by IsraelOrtuno** [comment]
Despite it seems to not be documented yet, Nuxt already has support for `scrollToTop` since RC-12 if I am not wrong:

```
definePageMeta({
  scrollToTop: true
})
```

**Comment by andorfermichael** [comment]
For me scrollTop is not working when using when changing route via [navigateTo](https://nuxt.com/docs/api/utils/navigate-to/)

**Comment by ipanamski** [comment]
In case anybody is using Vuetify 3, you can use this plugin
```
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook("page:finish", async (a) => {
    document.body.scrollTo({ top: 0 });
  });
});
```

**Comment by BboyAkers** [comment]
Update!!There's officially a scrollToTop property! 🙂 
https://nuxt.com/docs/migration/component-options#scrolltotop

**Reply by MartinX3** [reply]
How to use it with `<script setup>`?

**Reply by vanling** [reply]
That is nuxt2 documentation :)

**Reply by BboyAkers** [reply]
Sorry pasted the wrong link, updated!

**Reply by vanling** [reply]
Thank you. However its not quite there yet, thats why this topic is such a large 'discussion' ;)


> **This feature is not yet supported in Nuxt 3.** If you want to overwrite the default scroll behavior of vue-router, you can do so in ~/app/router.options.ts (see [docs](https://nuxt.com/docs/guide/directory-structure/pages/#router-options)) for more info.

**Reply by descovi** [reply]
I am forced to avoid using the nuxt link component because even with these solutions I see the scroll start before the page changes; I think it's unfortunate; do you know if nuxt has any plans to address this issue in future updates?

**Reply by danielroe** [reply]
This is an old discussion. If you are experiencing an issue, please open an issue with a reproduction 🙏

**Reply by descovi** [reply]
is there an official solution to the scrolltop issue?
because this discussion start with this information: https://nuxt.com/docs/migration/component-options#scrolltotop where is written that scrolltotop is not supported in nuxt 3

**Comment by jboada** [comment]
> is there an official solution to the scrolltop issue? because this discussion start with this information: https://nuxt.com/docs/migration/component-options#scrolltotop where is written that scrolltotop is not supported in nuxt 3

It seems to be by now.

I have been googling around and so far, the possible solutions are this post and this example [https://stackblitz.com/edit/nuxt-3-nuxtlink-scroll-to-page-anchor?file=pages%2Findex.vue,app%2Frouter.options.ts](https://stackblitz.com/edit/nuxt-3-nuxtlink-scroll-to-page-anchor?file=pages%2Findex.vue,app%2Frouter.options.ts)

**Reply by christine927t** [reply]
That stackblitz example works really well for me. Thanks for posting it!

**Reply by Kadae** [reply]
Timeout is a bad decision, because the user will see the page in the center or at the bottom 500 ms (as in example) before it becomes the top.

**Comment by bkan36** [comment]
My solution if it helps:
targetScroll is my id attribute
I wrapped it in a setTimeout because the code is triggered after button click
which grow a dynamic list.
```
        setTimeout(() => {
            const element = document.getElementById('targetScroll');
            element.scrollIntoView({ behavior: "smooth" })
        }, 100)
```

**Reply by Kadae** [reply]
Timeout is a bad decision, because the user will see the page in the center or at the bottom 100 ms before it becomes the top.

**Comment by armando-ota** [comment]
none of the above examples works ... #justSayin ... 

Neither RouterConfig in app/router.options.ts 
neither all the hacks provided here ... using nuxt ^3.8.2 localy and on netlify 
no luck :(

**Comment by EndlessWord** [comment]
1.modify node_modules\nuxt\dist\pages\runtime\plugins\router.js:40, delete scrollBehavior function
2.create app/router.options.ts
```
import type { RouterConfig } from '@nuxt/schema'

// https://router.vuejs.org/api/#routeroptions
export default <RouterConfig>{
  scrollBehavior: (to, from, savedPosition) => {
    // scroll to hash, useful for using to="#some-id" in NuxtLink
    // ex: <NuxtLink to="#top"> To Top </ NuxtLink>
    if (to.hash) {
      return {
        el: to.hash,
        behavior: 'smooth',
      }
    }

    // The remainder is optional but maybe useful as well

    // if link is to same page, scroll to top with smooth behavior
    if (to === from) {
      return {
        left: 0,
        top: 0,
        behavior: 'smooth',
      }
    }

    // this will use saved scroll position on browser forward/back navigation
    return false
  },
}
```
if use lenis,maybe this can help you
```
import type { RouterConfig } from '@nuxt/schema'

export default <RouterConfig>{
  scrollBehavior: (to, from) => {
    const nuxtApp = useNuxtApp()

    if (to.hash) {
      nuxtApp.$lenis.scrollTo(to.hash)
      return false
    }

    // The remainder is optional but maybe useful as well

    // if link is to same page, scroll to top with smooth behavior
    if (to === from) {
      nuxtApp.$lenis.scrollTo(0)
      return false
    }

    // this will use saved scroll position on browser forward/back navigation
    nuxtApp.$lenis.scrollTo(0, {
      immediate: true,
      force: true,
    })
    return false
  },
}

```

**Reply by armando-ota** [reply]
> 1.modify node_modules\nuxt\dist\pages\runtime\plugins\router.js:40, delete scrollBehavior function

can't just change this ... will not work on netlify deploy ... or will it ?

**Comment by uitlaber** [comment]
My solution  add to default layout root element ref  and 
```
const mainEl = useTemplateRef<HTMLElement>('mainEl')

watch(route, (to, from) => {
  if (import.meta.client) {
    setTimeout(() => {
      smooth.value = true
      mainEl.value?.scrollTo({
        top: 0,
        behavior: 'auto',
      })
    }, 100)
  }
})


  <div
    ref="mainEl"
    class="main-wrapper"
  > </div>
```

**Reply by OrbisK** [reply]
```ts
if (import.meta.client) {
  watch(route, (to, from) => {
    setTimeout(() => {
      smooth.value = true
      mainEl.value?.scrollTo({
        top: 0,
        behavior: 'auto',
      })
    }, 100)
  })
}
```
might be slightly better :)

**Comment by snakysnake** [comment]
plugins/scroll-top.client.ts
```
export default defineNuxtPlugin(() => {
	const router = useRouter();

	// Function to scroll to top
	const scrollTop = () => {
		window.scrollTo(0, 0);
	};

	// Watch for route changes
	router.afterEach((to, from) => {
		// Only scroll to top if:
		// 1. It's a different page (not just anchor navigation)
		// 2. The pathname has changed (not just hash change)
		if (to.path !== from.path) {
			// Use nextTick to ensure the new page has rendered
			nextTick(() => {
				scrollTop();
			});
		}
	});
});
```

**Comment by samuelreichor** [comment]
Honestly I hate how complicated such a "simple" task can be.. I would love to see a real fix for that.

**Comment by dngraphisme** [comment]
Hello,

I've always had this problem, and no proper fix solves it. It's always a random jump across all browsers...

Even with Lenis, no matter how much I try stopping it, blocking it, adding an extra router.option, or even managing the scroll in the transition, nothing works.
